### PR TITLE
Fix crash rx single

### LINKED ISSUE #3123: Bug? @user:server.tld is already in the room when creating a new direct message with the user themselves/himself/herself/self/oneself
#### Describe the bug

Not sure if a bug or a feature... 😅 

#### To Reproduce

1. Open authenticated and unlocked Element Android
2. Click on new direct message FAB
3. Select the current user, i.e. oneself
4. Click CREATE
5. Creating room...
6. Observe warning/info/error dialog: `@user:matrix.tld is already in the room.`

#### Expected behavior

The same behaviour as on the Element web client, or the other way around: the web should show the same behaviour as in the Element Android client. On the web an 'Empty room' with just the user is created when trying to create a direct message with oneself.

#### Smartphone (please complete the following information):
 - Device: Google Pixel 3a
 - OS: Android 11

#### Additional context
 - App version and store: 1.1.3 [40101032] (G-b1729)
 - Homeserver: matrix.org

### LINKED ISSUE #3180: Crash with `rxSingle`
Fails fast feature must be on to see this problem. To reproduce: try to create a DM and type some chars in the search user field, in order to cancel a request.

```
    io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | Cancelled(throwable=kotlinx.coroutines.JobCancellationException: RxSingleCoroutine was cancelled; job=RxSingleCoroutine{Cancelling}@8b85a04)
        at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:9)
        at io.reactivex.plugins.RxJavaPlugins.handleUndeliverableException(RxJavaPlugins.java:2)
        at kotlinx.coroutines.rx2.RxSingleCoroutine.onCancelled(RxSingle.kt:3)
        at kotlinx.coroutines.AbstractCoroutine.onCompletionInternal(AbstractCoroutine.kt:2)
        at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:17)
        at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:27)
        at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:2)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:2)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:10)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:14)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:10)
     Caused by: Cancelled(throwable=kotlinx.coroutines.JobCancellationException: RxSingleCoroutine was cancelled; job=RxSingleCoroutine{Cancelling}@8b85a04)
        at org.matrix.android.sdk.internal.session.user.model.DefaultSearchUserTask.execute(SearchUserTask.kt:36)
        at org.matrix.android.sdk.internal.session.user.model.DefaultSearchUserTask$execute$1.invokeSuspend(Unknown Source:12)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:3)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:14) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:10) 
     Caused by: kotlinx.coroutines.JobCancellationException: RxSingleCoroutine was cancelled; job=RxSingleCoroutine{Cancelling}@8b85a04
```

### PR DESCRIPTION
**Pull request by bmarty**
Fix a crash in rxSingle, and also a problem when creating a DM inviting the current user.

Fixes #3180 (Do not wrap CancellationException to Failure.Cancelled, else RxCancellable will throw CC @Dominaezzz :) )
Fixes #3123

Also try to reduce the usage of Rx (the easy part), which will at the end be removed.

### CODE REVIEW

File: vector/src/main/java/im/vector/app/core/mvrx/ResultExtension.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package im.vector.app.core.mvrx
+
+import com.airbnb.mvrx.Async
+import com.airbnb.mvrx.Fail
+import com.airbnb.mvrx.Success
+
+fun <A> Result<A>.foldToAsync(): Async<A> = fold(
```
Comment (Dominaezzz) [inline vector/src/main/java/im/vector/app/core/mvrx/ResultExtension.kt:23]: Seems this is always called after `runCatching`. Might be worth making a function that does both `runCatching` and `foldToAsync` in one go.

Comment (ganfra) [reply vector/src/main/java/im/vector/app/core/mvrx/ResultExtension.kt:23]: Also, be aware that we are losing the Loading state here. FTR, Mavericks 2 has full coroutine support and already provide methods for this case.

Comment (bmarty) [reply vector/src/main/java/im/vector/app/core/mvrx/ResultExtension.kt]: Thanks, I've handled your remarks.

**Review by BillCarsonFr — APPROVED** [review]
LGTM

**Pull request merged**